File: mv-design-pro/backend/src/api/switchgear_config.py
"""
Switchgear Config API — CRUD + walidacja konfiguracji rozdzielnicy.

RUN #3I COMMIT 2: REST API dla SwitchgearConfigV1.

Endpointy:
- GET  /api/switchgear/{station_id}/config         → pobranie konfiguracji
- PUT  /api/switchgear/{station_id}/config         → zapis konfiguracji
- POST /api/switchgear/{station_id}/validate       → walidacja (issues + fixActions)

Wzorzec: in-memory store (jak sld_overrides.py), Pydantic na granicy API.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel

from domain.switchgear_config import (
    CatalogBindingV1,
    DeviceConfigV1,
    FieldConfigV1,
    ProtectionBindingV1,
    SwitchgearConfigV1,
    canonicalize_config,
    compute_config_hash,
    validate_switchgear_config,
)
from domain.field_device import (
    AparatTypeV1,
    DeviceTypeV1,
    FieldRoleV1,
    PoleTypeV1,
)

router = APIRouter(
    prefix="/api/switchgear",
    tags=["switchgear-config"],
)

# =============================================================================
# REQUEST / RESPONSE MODELS (Pydantic — API boundary only)
# =============================================================================


class FieldConfigRequest(BaseModel):
    field_id: str
    pole_type: str
    field_role: str
    bus_section_id: str | None = None


class DeviceConfigRequest(BaseModel):
    device_id: str
    field_id: str
    device_type: str
    aparat_type: str


class CatalogBindingRequest(BaseModel):
    device_id: str
    catalog_id: str
    catalog_name: str
    manufacturer: str | None = None
    catalog_version: str | None = None


class ProtectionBindingRequest(BaseModel):
    relay_device_id: str
    cb_device_id: str


class PutConfigRequest(BaseModel):
    fields: list[FieldConfigRequest] = []
    devices: list[DeviceConfigRequest] = []
    catalog_bindings: list[CatalogBindingRequest] = []
    protection_bindings: list[ProtectionBindingRequest] = []


class ConfigResponse(BaseModel):
    config_version: str
    station_id: str
    fields: list[dict[str, Any]]
    devices: list[dict[str, Any]]
    catalog_bindings: list[dict[str, Any]]
    protection_bindings: list[dict[str, Any]]
    canonical_hash: str


class ValidationIssueResponse(BaseModel):
    code: str
    severity: str
    message_pl: str
    element_id: str | None = None
    field_id: str | None = None
    device_id: str | None = None


class FixActionResponse(BaseModel):
    code: str
    action: str
    message_pl: str
    station_id: str
    field_id: str | None = None
    device_id: str | None = None


class ValidateResponse(BaseModel):
    valid: bool
    issues: list[ValidationIssueResponse]
    fix_actions: list[FixActionResponse]
    canonical_hash: str


# =============================================================================
# IN-MEMORY STORE
# =============================================================================

_config_store: dict[str, SwitchgearConfigV1] = {}


def _get_config(station_id: str) -> SwitchgearConfigV1:
    """Pobierz konfiguracje lub zwroc pusta."""
    return _config_store.get(
        station_id,
        SwitchgearConfigV1(station_id=station_id),
    )


def _parse_config(station_id: str, req: PutConfigRequest) -> SwitchgearConfigV1:
    """Parse request into domain model."""
    fields = tuple(
        FieldConfigV1(
            field_id=f.field_id,
            pole_type=PoleTypeV1(f.pole_type),
            field_role=FieldRoleV1(f.field_role),
            bus_section_id=f.bus_section_id,
        )
        for f in req.fields
    )
    devices = tuple(
        DeviceConfigV1(
            device_id=d.device_id,
            field_id=d.field_id,
            device_type=DeviceTypeV1(d.device_type),
            aparat_type=AparatTypeV1(d.aparat_type),
        )
        for d in req.devices
    )
    catalog_bindings = tuple(
        CatalogBindingV1(
            device_id=b.device_id,
            catalog_id=b.catalog_id,
            catalog_name=b.catalog_name,
            manufacturer=b.manufacturer,
            catalog_version=b.catalog_version,
        )
        for b in req.catalog_bindings
    )
    protection_bindings = tuple(
        ProtectionBindingV1(
            relay_device_id=p.relay_device_id,
            cb_device_id=p.cb_device_id,
        )
        for p in req.protection_bindings
    )
    return SwitchgearConfigV1(
        station_id=station_id,
        fields=fields,
        devices=devices,
        catalog_bindings=catalog_bindings,
        protection_bindings=protection_bindings,
    )


def _to_response(config: SwitchgearConfigV1) -> ConfigResponse:
    """Convert domain model to API response."""
    canonical = canonicalize_config(config)
    return ConfigResponse(
        config_version=canonical.config_version,
        station_id=canonical.station_id,
        fields=[f.to_dict() for f in canonical.fields],
        devices=[d.to_dict() for d in canonical.devices],
        catalog_bindings=[b.to_dict() for b in canonical.catalog_bindings],
        protection_bindings=[p.to_dict() for p in canonical.protection_bindings],
        canonical_hash=compute_config_hash(canonical),
    )


# =============================================================================
# ENDPOINTS
# =============================================================================


@router.get(
    "/{station_id}/config",
    response_model=ConfigResponse,
    summary="Pobierz konfiguracje rozdzielnicy",
)
async def get_config(station_id: str) -> ConfigResponse:
    """Pobierz konfiguracje rozdzielnicy (kanonizowana, z hashem)."""
    config = _get_config(station_id)
    return _to_response(config)


@router.put(
    "/{station_id}/config",
    response_model=ConfigResponse,
    summary="Zapisz konfiguracje rozdzielnicy",
)
async def put_config(station_id: str, req: PutConfigRequest) -> ConfigResponse:
    """Zapisz konfiguracje rozdzielnicy (kanonizowana, deterministyczna)."""
    config = _parse_config(station_id, req)
    canonical = canonicalize_config(config)
    _config_store[station_id] = canonical
    return _to_response(canonical)


@router.post(
    "/{station_id}/validate",
    response_model=ValidateResponse,
    summary="Waliduj konfiguracje rozdzielnicy",
)
async def validate_config(
    station_id: str,
    req: PutConfigRequest,
) -> ValidateResponse:
    """Waliduj konfiguracje bez zapisu. Zwraca issues + fixActions."""
    config = _parse_config(station_id, req)
    result = validate_switchgear_config(config)
    canonical = canonicalize_config(config)
    return ValidateResponse(
        valid=result.valid,
        issues=[
            ValidationIssueResponse(
                code=i.code,
                severity=i.severity.value,
                message_pl=i.message_pl,
                element_id=i.element_id,
                field_id=i.field_id,
                device_id=i.device_id,
            )
            for i in result.issues
        ],
        fix_actions=[
            FixActionResponse(
                code=fa.code,
                action=fa.action.value,
                message_pl=fa.message_pl,
                station_id=fa.station_id,
                field_id=fa.field_id,
                device_id=fa.device_id,
            )
            for fa in result.fix_actions
        ],
        canonical_hash=compute_config_hash(canonical),
    )

File: mv-design-pro/frontend/src/ui/sld/core/switchgearConfig.ts
/**
 * SwitchgearConfigV1 — Konfiguracja rozdzielnicy (pola/aparaty/katalogi/ochrona).
 *
 * RUN #3I COMMIT 1: Kontrakt TS 1:1 z backendem (domain/switchgear_config.py).
 *
 * CANONICAL CONTRACT (BINDING):
 * - Immutable (readonly).
 * - Deterministic: sortowanie po id, kanoniczny hash (FNV-1a), permutation invariant.
 * - ZAKAZ auto-uzupelnien: brak danych -> FixAction (stabilny kod PL).
 * - ZAKAZ domyslnych parametrow urzadzen — brak = ValidationIssue + FixAction.
 * - PV/BESS zawsze przez transformator (SN/nN lub blokowy).
 * - catalogRef wymagany dla kazdego aparatu — brak -> CATALOG_REF_MISSING.
 *
 * ALIGNMENT:
 * - backend: domain/switchgear_config.py
 * - frontend: fieldDeviceContracts.ts (PoleTypeV1, AparatTypeV1, FieldRoleV1, DeviceTypeV1)
 */

import {
  type DeviceTypeV1,
  DeviceTypeV1 as DT,
  type FieldRoleV1,
  FieldRoleV1 as FR,
  type PoleTypeV1,
  type AparatTypeV1,
} from './fieldDeviceContracts';

// =============================================================================
// VERSION
// =============================================================================

export const SWITCHGEAR_CONFIG_VERSION = '1.0' as const;

// =============================================================================
// CONFIG ISSUE SEVERITY
// =============================================================================

export const ConfigIssueSeverity = {
  BLOCKER: 'BLOCKER',
  WARNING: 'WARNING',
} as const;

export type ConfigIssueSeverity =
  (typeof ConfigIssueSeverity)[keyof typeof ConfigIssueSeverity];

// =============================================================================
// VALIDATION ISSUE
// =============================================================================

export interface ConfigValidationIssueV1 {
  readonly code: string;
  readonly severity: ConfigIssueSeverity;
  readonly messagePl: string;
  readonly elementId: string | null;
  readonly fieldId: string | null;
  readonly deviceId: string | null;
}

// =============================================================================
// FIX ACTION TYPE
// =============================================================================

export const FixActionType = {
  NAVIGATE_TO_WIZARD_FIELD: 'NAVIGATE_TO_WIZARD_FIELD',
  NAVIGATE_TO_WIZARD_DEVICE: 'NAVIGATE_TO_WIZARD_DEVICE',
  NAVIGATE_TO_WIZARD_CATALOG_PICKER: 'NAVIGATE_TO_WIZARD_CATALOG_PICKER',
  NAVIGATE_TO_WIZARD_PROTECTION: 'NAVIGATE_TO_WIZARD_PROTECTION',
} as const;

export type FixActionType =
  (typeof FixActionType)[keyof typeof FixActionType];

// =============================================================================
// FIX ACTION
// =============================================================================

export interface ConfigFixActionV1 {
  readonly code: string;
  readonly action: FixActionType;
  readonly messagePl: string;
  readonly stationId: string;
  readonly fieldId: string | null;
  readonly deviceId: string | null;
}

// =============================================================================
// CATALOG BINDING V1
// =============================================================================

export interface CatalogBindingV1 {
  readonly deviceId: string;
  readonly catalogId: string;
  readonly catalogName: string;
  readonly manufacturer: string | null;
  readonly catalogVersion: string | null;
}

// =============================================================================
// PROTECTION BINDING V1
// =============================================================================

export interface ProtectionBindingV1 {
  readonly relayDeviceId: string;
  readonly cbDeviceId: string;
}

// =============================================================================
// DEVICE CONFIG V1
// =============================================================================

export interface DeviceConfigV1 {
  readonly deviceId: string;
  readonly fieldId: string;
  readonly deviceType: DeviceTypeV1;
  readonly aparatType: AparatTypeV1;
}

// =============================================================================
// FIELD CONFIG V1
// =============================================================================

export interface FieldConfigV1 {
  readonly fieldId: string;
  readonly poleType: PoleTypeV1;
  readonly fieldRole: FieldRoleV1;
  readonly busSectionId: string | null;
}

// =============================================================================
// SWITCHGEAR CONFIG V1
// =============================================================================

export interface SwitchgearConfigV1 {
  readonly configVersion: string;
  readonly stationId: string;
  readonly fields: readonly FieldConfigV1[];
  readonly devices: readonly DeviceConfigV1[];
  readonly catalogBindings: readonly CatalogBindingV1[];
  readonly protectionBindings: readonly ProtectionBindingV1[];
}

// =============================================================================
// VALIDATION RESULT
// =============================================================================

export interface SwitchgearConfigValidationResultV1 {
  readonly valid: boolean;
  readonly issues: readonly ConfigValidationIssueV1[];
  readonly fixActions: readonly ConfigFixActionV1[];
}

// =============================================================================
// VALIDATION CODES (stable PL — 1:1 with backend)
// =============================================================================

export const SwitchgearConfigValidationCode = {
  CATALOG_REF_MISSING: 'catalog.ref_missing',
  FIELD_MISSING_REQUIRED_DEVICE: 'field.missing_required_device',
  DEVICE_MISSING_REQUIRED_PARAMETER: 'device.missing_required_parameter',
  PROTECTION_BINDING_MISSING: 'protection.binding_missing',
  PV_BESS_TRANSFORMER_MISSING: 'pv_bess.transformer_missing',
  FIELD_DUPLICATE_ID: 'field.duplicate_id',
  DEVICE_DUPLICATE_ID: 'device.duplicate_id',
  DEVICE_ORPHAN: 'device.orphan_no_field',
  CATALOG_BINDING_ORPHAN: 'catalog_binding.orphan_no_device',
  PROTECTION_BINDING_ORPHAN: 'protection_binding.orphan_no_device',
} as const;

export type SwitchgearConfigValidationCode =
  (typeof SwitchgearConfigValidationCode)[keyof typeof SwitchgearConfigValidationCode];

// =============================================================================
// REQUIRED DEVICES per FieldRole (1:1 with backend)
// =============================================================================

export const REQUIRED_DEVICES: Record<FieldRoleV1, readonly DeviceTypeV1[]> = {
  [FR.LINE_IN]: [DT.CB, DT.CABLE_HEAD],
  [FR.LINE_OUT]: [DT.CB, DT.CABLE_HEAD],
  [FR.LINE_BRANCH]: [DT.CB, DT.CABLE_HEAD],
  [FR.TRANSFORMER_SN_NN]: [DT.CB, DT.CT, DT.RELAY, DT.TRANSFORMER_DEVICE, DT.CABLE_HEAD],
  [FR.PV_SN]: [DT.CB, DT.CT, DT.RELAY, DT.GENERATOR_PV, DT.CABLE_HEAD],
  [FR.BESS_SN]: [DT.CB, DT.CT, DT.RELAY, DT.GENERATOR_BESS, DT.CABLE_HEAD],
  [FR.COUPLER_SN]: [DT.CB],
  [FR.BUS_TIE]: [DT.CB],
  [FR.MAIN_NN]: [DT.ACB],
  [FR.FEEDER_NN]: [DT.FUSE],
  [FR.PV_NN]: [DT.ACB, DT.CT, DT.RELAY, DT.GENERATOR_PV],
  [FR.BESS_NN]: [DT.ACB, DT.CT, DT.RELAY, DT.GENERATOR_BESS],
};

// PV/BESS SN field roles that require transformer
export const PV_BESS_SN_ROLES: ReadonlySet<FieldRoleV1> = new Set([
  FR.PV_SN,
  FR.BESS_SN,
]);

// =============================================================================
// CANONICAL SERIALIZATION
// =============================================================================

/**
 * Kanonizuje konfiguracje — sortuje deterministycznie.
 *
 * Sortowanie:
 * - fields: po fieldId
 * - devices: po deviceId
 * - catalogBindings: po deviceId
 * - protectionBindings: po relayDeviceId, cbDeviceId
 */
export function canonicalizeConfig(config: SwitchgearConfigV1): SwitchgearConfigV1 {
  return {
    configVersion: config.configVersion,
    stationId: config.stationId,
    fields: [...config.fields].sort((a, b) => a.fieldId.localeCompare(b.fieldId)),
    devices: [...config.devices].sort((a, b) => a.deviceId.localeCompare(b.deviceId)),
    catalogBindings: [...config.catalogBindings].sort((a, b) =>
      a.deviceId.localeCompare(b.deviceId),
    ),
    protectionBindings: [...config.protectionBindings].sort((a, b) => {
      const cmp = a.relayDeviceId.localeCompare(b.relayDeviceId);
      return cmp !== 0 ? cmp : a.cbDeviceId.localeCompare(b.cbDeviceId);
    }),
  };
}

/**
 * Serializuje konfiguracje do kanonicznego JSON (klucze posortowane).
 *
 * Format identyczny z backendem: sort_keys=True, separators=(",", ":").
 */
function configToCanonicalJson(config: SwitchgearConfigV1): string {
  const canonical = canonicalizeConfig(config);
  const data = {
    catalog_bindings: canonical.catalogBindings.map(b => ({
      catalog_id: b.catalogId,
      catalog_name: b.catalogName,
      catalog_version: b.catalogVersion,
      device_id: b.deviceId,
      manufacturer: b.manufacturer,
    })),
    config_version: canonical.configVersion,
    devices: canonical.devices.map(d => ({
      aparat_type: d.aparatType,
      device_id: d.deviceId,
      device_type: d.deviceType,
      field_id: d.fieldId,
    })),
    fields: canonical.fields.map(f => ({
      bus_section_id: f.busSectionId,
      field_id: f.fieldId,
      field_role: f.fieldRole,
      pole_type: f.poleType,
    })),
    protection_bindings: canonical.protectionBindings.map(p => ({
      cb_device_id: p.cbDeviceId,
      relay_device_id: p.relayDeviceId,
    })),
    station_id: canonical.stationId,
  };
  return JSON.stringify(data);
}

// =============================================================================
// HASHING (FNV-1a 32-bit — consistent with other SLD contracts)
// =============================================================================

/**
 * FNV-1a 32-bit hash — identyczny algorytm jak w visualGraph, geometryOverrides.
 */
function fnv1a32(input: string): string {
  let hash = 0x811c9dc5;
  for (let i = 0; i < input.length; i++) {
    hash ^= input.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  return (hash >>> 0).toString(16).padStart(8, '0');
}

/**
 * Oblicza deterministyczny hash konfiguracji.
 *
 * Permutation invariant: kolejnosc pol/aparatow nie wplywa na hash.
 * Kanoniczny JSON z posortowanymi kluczami.
 */
export function computeConfigHash(config: SwitchgearConfigV1): string {
  const json = configToCanonicalJson(config);
  return fnv1a32(json);
}

// =============================================================================
// EMPTY CONFIG FACTORY
// =============================================================================

export function emptyConfig(stationId: string): SwitchgearConfigV1 {
  return {
    configVersion: SWITCHGEAR_CONFIG_VERSION,
    stationId,
    fields: [],
    devices: [],
    catalogBindings: [],
    protectionBindings: [],
  };
}
